src/containers/Article.js

```diff
@@ -54,32 +65,115 @@ class Article extends PureComponent {
     this.handleFontLevelChange = this._handleFontLevelChange.bind(this)
     this.onToggleTabExpanded = this._onToggleTabExpanded.bind(this)
     this._articleBody = React.createRef()
-    this.timerId = React.createRef()
+    this.readingCountTimerId = React.createRef()
+    this.inactiveTimerId = React.createRef()
     this.state = {
       isExpanded: false,
       isReachedArticleReadTargetHeight: false,
       isReachedArticleReadTargetTime: false,
+      isBeenRead: false,
+      startReadingTime: 0,
+      isActive: true,
+      activeTime: 0,
     }
-    this.handleScroll = this._handleScroll.bind(this)
+    this.handleScroll = _.debounce(this._handleScroll.bind(this), 500)
+    this.handleVisibilityChange = _.throttle(
+      this._handleVisibilityChange.bind(this),
+      1000
+    )
+    this.handleUserActivity = _.debounce(
+      this._handleUserActivity.bind(this),
+      500
+    )
   }
 
-  startTimer() {
-    if (this.timerId.current) {
-      window.clearTimeout(this.timerId.current)
+  startReadingCountTimer() {
+    if (this.readingCountTimerId.current) {
+      window.clearTimeout(this.readingCountTimerId.current)
     }
-    this.timerId.current = window.setTimeout(() => {
+    this.readingCountTimerId.current = window.setTimeout(() => {
       this.setState({ isReachedArticleReadTargetTime: true })
     }, articleReadCountConditionConfig.reading_time)
   }
 
+  calculateActiveTime() {
+    const elapsedTime = Math.floor(Date.now() - this.state.startReadingTime)
+    this.setState(prevState => ({
+      activeTime: prevState.activeTime + elapsedTime,
+      startReadingTime: Date.now(),
+    }))
+  }
+
+  startInactiveTimer() {
+    if (this.inactiveTimerId.current) {
+      window.clearTimeout(this.inactiveTimerId.current)
+    }
+    this.inactiveTimerId.current = window.setTimeout(() => {
+      this.calculateActiveTime()
+      this.setState({
+        isActive: false,
+      })
+    }, articleReadTimeConditionConfig.inactive_time)
+  }
+
+  sendReadCount() {
+    const { jwt, userID, postID, setUserAnalyticsData } = this.props
+    setUserAnalyticsData(jwt, userID, postID, { readPostCount: true })
+  }
+
+  sendActiveTime() {
+    const { jwt, userID, postID, setUserAnalyticsData } = this.props
+    const { activeTime } = this.state
+    const activeSec = Math.round(activeTime / 1000)
+    setUserAnalyticsData(jwt, userID, postID, { readPostSec: activeSec })
+    this.setState({
+      activeTime: 0,
+    })
+  }
+
+  _handleVisibilityChange() {
+    if (document.visibilityState === 'hidden') {
+      if (
+        this.state.activeTime >= articleReadTimeConditionConfig.min_active_time
+      ) {
+        this.sendActiveTime()
```

Comment:
Question: should we call `calculateActiveTime` before `sendActiveTime` here ?

Reply:
fixed